cpp/src/text/edit_distance.cu

```diff
@@ -137,154 +129,157 @@ struct edit_distance_matrix_levenshtein_algorithm {
-    strings_count * strings_count,
-    [d_strings, d_offsets, strings_count] __device__(cudf::size_type idx) {
-      auto const row = idx / strings_count;
-      auto const col = idx % strings_count;
-      if (row >= col) return;  // compute only the top half
-      cudf::string_view const d_str1 =
-        d_strings.is_null(row) ? cudf::string_view{} : d_strings.element<cudf::string_view>(row);
-      cudf::string_view const d_str2 =
-        d_strings.is_null(col) ? cudf::string_view{} : d_strings.element<cudf::string_view>(col);
-      if (d_str1.empty() || d_str2.empty()) { return; }
-      // the temp size needed is 2 integers per character of the shorter string
-      d_offsets[idx - ((row + 1) * (row + 2)) / 2] =
-        (cuda::std::min(d_str1.length(), d_str2.length()) + 1) * 2;
-    });
+  auto const n_upper     = (input.size() * (input.size() - 1L)) / 2L;
+  auto const output_size = input.size() * input.size();
+  rmm::device_uvector<std::ptrdiff_t> offsets(n_upper + 1, stream);
+  thrust::uninitialized_fill(rmm::exec_policy_nosync(stream), offsets.begin(), offsets.end(), 0);
+  thrust::for_each_n(rmm::exec_policy_nosync(stream),
+                     thrust::counting_iterator<cudf::size_type>(0),
+                     output_size,
+                     calculate_matrix_compute_buffer_fn{*d_strings, offsets.data()});
 
   // get the total size for the compute buffer
-  int64_t compute_size =
-    thrust::reduce(rmm::exec_policy(stream), offsets.begin(), offsets.end(), int64_t{0});
-  // convert sizes to offsets in-place
-  thrust::exclusive_scan(rmm::exec_policy(stream), offsets.begin(), offsets.end(), offsets.begin());
+  // and convert sizes to offsets in-place
+  auto const compute_size =
+    cudf::detail::sizes_to_offsets(offsets.begin(), offsets.end(), offsets.begin(), 0, stream);
+
   // create the compute buffer
   rmm::device_uvector<cudf::size_type> compute_buffer(compute_size, stream);
   auto d_buffer = compute_buffer.data();
 
   // compute the edit distance into the output column
-  auto results = cudf::make_fixed_width_column(cudf::data_type{cudf::type_to_id<cudf::size_type>()},
-                                               strings_count * strings_count,
-                                               rmm::device_buffer{0, stream, mr},
-                                               0,
-                                               stream,
-                                               mr);
+  auto results = cudf::make_fixed_width_column(
+    output_type, output_size, rmm::device_buffer{0, stream, mr}, 0, stream, mr);
   auto d_results = results->mutable_view().data<cudf::size_type>();
   thrust::for_each_n(
-    rmm::exec_policy(stream),
-    thrust::make_counting_iterator<cudf::size_type>(0),
-    strings_count * strings_count,
-    edit_distance_matrix_levenshtein_algorithm{d_strings, d_buffer, d_offsets, d_results});
+    rmm::exec_policy_nosync(stream),
```

Comment:
Pardon my ignorance, but when is it appropriate to switch to the `_nosync` exec policy?  Is it safe here because these calls are on the same stream?

Reply:
Appropriate for `thrust` APIs that do not return anything or do not require a sync to build the result.
Some `thrust` calls return a value (e.g. `reduce`) or an iterator (e.g. `copy_if`) that require a sync to produce a host value. There are others that return values that do not require a sync (e.g. `copy`) or that we do not need the result.